arbnode/message_pruner.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package arbnode
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/offchainlabs/nitro/staker"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+
+	flag "github.com/spf13/pflag"
+)
+
+type MessagePruner struct {
+	stopwaiter.StopWaiter
+	transactionStreamer *TransactionStreamer
+	inboxTracker        *InboxTracker
+	staker              *staker.Staker
+	config              MessagePrunerConfigFetcher
+}
+
+type MessagePrunerConfig struct {
+	Enable               bool          `koanf:"enable"`
+	MessagePruneInterval time.Duration `koanf:"prune-interval" reload:"hot"`
+}
+
+type MessagePrunerConfigFetcher func() *MessagePrunerConfig
+
+var DefaultMessagePrunerConfig = MessagePrunerConfig{
+	Enable:               false,
+	MessagePruneInterval: time.Minute,
+}
+
+func MessagePrunerConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Bool(prefix+".enable", DefaultMessagePrunerConfig.Enable, "enable message pruning")
+	f.Duration(prefix+".prune-interval", DefaultMessagePrunerConfig.MessagePruneInterval, "interval for running message pruner")
+}
+
+func NewMessagePruner(transactionStreamer *TransactionStreamer, inboxTracker *InboxTracker, staker *staker.Staker, config MessagePrunerConfigFetcher) *MessagePruner {
+	return &MessagePruner{
+		transactionStreamer: transactionStreamer,
+		inboxTracker:        inboxTracker,
+		staker:              staker,
+		config:              config,
+	}
+}
+
+func (m *MessagePruner) Start(ctxIn context.Context) {
+	m.CallIteratively(func(ctx context.Context) time.Duration {
+		m.prune(ctx)
```

Comment:
prune can return MessagePruneInterval and you won't need this extra function.. but not really important

Reply:
done.